Modifications to build profile documentation

SenRamakri opened the pull request:
Modifications to build profile documentation to explicitly capture the fact that this is build-profile and not just any profile. This helps us create other profiles for other purposes like bare-metal.

SenRamakri committed:
Modifications to build profile documentation to be explicit about the fact that this is build-profile and not just any profile. This helps us create other profiles for other purposes like bare-metal

Commit:
Edit updates to build_profiles.md

Edit updates, mostly for grammar and style.

AnotherButler commented on docs/tools/CLI/build_profiles.md:
```diff
@@ -1,42 +1,42 @@
+   - Breaks the local file system on the [Mbed interface](../introduction/index.html) on some boards.
 
-### User defined profiles
+### User-defined
```
User-defined what?

SenRamakri replied:
Fixed by adding "build profile", please review.

SenRamakri committed:
Review comment fixes

Commit:
Edit heading change to build_profiles.md

Remove code format from heading for better rendering and UX.

AnotherButler reviewed (approved):
Thanks for the PR 👍

theotherjimmy commented on docs/tools/CLI/build_profiles.md:
```diff
@@ -1,42 +1,42 @@
 <h2 id="build-profiles">Build profiles</h2>
 
-Arm Mbed OS 5 supports three primary build profiles: *develop*, *debug* and *release*. The Online Compiler uses the *develop* profile. When building from Arm Mbed CLI, you can select a profile by adding the `--profile <profile>` flag. You can specify custom user-defined profiles by giving the path to the profile.
+Arm Mbed OS 5 defines three build configurations, each of which is known as a `build profile`. These are collections of toolchain flags used during the build. The three build profiles are *develop*, *debug* and *release*. The Mbed Online Compiler uses the *develop* build profile. When  building from Arm Mbed CLI, you can select the build configuration by adding the `--profile <`**build profile**  ...
```
Why use the term "build configuration" only once to define "build profile". This seems like it would only add to the confusion.

theotherjimmy replied:
```suggestion
Arm Mbed OS 5 defines three collections of toolchain flags used during the build. These are build profiles. The three build profiles are *develop*, *debug* and *release*. The Mbed Online Compiler uses the *develop* build profile. When  building from Arm Mbed CLI, you can select the build configuration by adding the `--profile <`**build profile** `name>` flag. You can also specify custom- or user-defined configurations by giving the path to the JSON file defining the build configuration.
```

theotherjimmy commented on docs/tools/CLI/build_profiles.md:
```diff
@@ -1,42 +1,42 @@
+The JSON object that represents a toolchain configuration is a dictionary mapping from toolchains, such as `GCC_ARM`, to their flags, such as `-O3`.
 
-The structure is as follows: each toolchain supported by a toolchain profile has a dictionary in the root dictionary. This dictionary contains a mapping from a flag type to a list of flags that should be passed to the corresponding part of the compiler suite.
+The structure is as follows: each toolchain to be supported has a dictionary in the root dictionary. This dictionary contains a mapping from a flag type to a list of flags that should be passed to the corresponding part of the compiler suite.
```
could we change all "dictionary" to "object" to match the JSON spec?

SenRamakri commented:
@theotherjimmy - Regarding "build configuration" we want to capture the fact that "build profile" is defined to be a set of configuration or toolchain flags. What's the best way to capture that? Do you have any suggestions?

theotherjimmy commented:
@SenRamakri using the word "configuration" here would be misleading, as a build profile is a set of toolchain flags.

Combining with the next sentence place the definition before the term, I recommend:
> Arm Mbed OS 5 defines three collections of toolchain flags used during the build, known as a `build profile`. ...

I'm not sure on the pluralization of the above sentence though. @AnotherButler knows.

AnotherButler commented:
Perhaps "Arm Mbed OS 5 defines three collections of toolchain flags used during the build. These are `build profiles`."? Can we make it plural and still keep it in code formatting?

theotherjimmy commented:
@AnotherButler We should not be using code formatting for build profiles. It's not something you type into the command line or source code.

theotherjimmy commented on docs/tools/CLI/build_profiles.md:
```diff
@@ -1,42 +1,42 @@
+### User-defined build profile
 
-A toolchain or build system profile is a set of flags that is guaranteed to be passed to the underlying compiler suite.
+As mentioned above, the `build profile` defines the set of flags that is guaranteed to be passed to the underlying compiler suite.
```
```suggestion
As mentioned above, the build profile defines the set of flags that is guaranteed to be passed to the underlying compiler suite.
```

theotherjimmy commented on docs/tools/CLI/build_profiles.md:
```diff
@@ -50,7 +50,7 @@ The required flag types are:
 #### Example
 
-An example of a toolchain profile:
+An example of a `build profile`:
```
```suggestion
An example of a build profile:
```

theotherjimmy reviewed (changes requested):
Please remove code formatting for "build profiles" as it's not something that's passed on the command line.

SenRamakri committed:
Review comments fixed

SenRamakri commented:
@theotherjimmy @AnotherButler - I have pushed a new commit with review comments fixes, please review.

Commit:
Edit changes to build_profiles.md

Edit some existing content that could be better, and some agreement in the most recent change.

AnotherButler commented on docs/tools/CLI/build_profiles.md:
```diff
@@ -1,42 +1,42 @@
+The JSON object that represents toolchain configurations for mapping each supported toolchains, such as `GCC_ARM`, to their flags, like `-O3`.
 
-The structure is as follows: each toolchain supported by a toolchain profile has a dictionary in the root dictionary. This dictionary contains a mapping from a flag type to a list of flags that should be passed to the corresponding part of the compiler suite.
+Each toolchain to be supported has an object in the root object. This object contains a mapping from a flag type to a list of flags that should be passed to the corresponding part of the compiler suite.
```
Does "This object" in the second sentence refer to the root object or the object in the root object? Please clarify.

AnotherButler commented on docs/tools/CLI/build_profiles.md:
```diff
@@ -1,42 +1,42 @@
+#### JSON build profile format
 
-The JSON object that represents a toolchain profile is a dictionary mapping from toolchains, such as `GCC_ARM`, to their flags, such as `-O3`.
+The JSON object that represents toolchain configurations for mapping each supported toolchains, such as `GCC_ARM`, to their flags, like `-O3`.
```
I think this sentence lost a subject. Also, "like" should probably be "such as", unless it can't be -03. ("Like" means similar to but not including, and "such as" means similar to and including.)

SenRamakri replied:
@AnotherButler - I made some changes to clarify the questions, please review again.

AnotherButler reviewed (changes requested):
I left two comments requesting two minor fixes.

SenRamakri committed:
Fixes for json root/child object explanations

AnotherButler commented on docs/tools/CLI/build_profiles.md:
```diff
@@ -1,42 +1,42 @@
+#### JSON build profile format
 
-The JSON object that represents a toolchain profile is a dictionary mapping from toolchains, such as `GCC_ARM`, to their flags, such as `-O3`.
+The JSON object that represents toolchain configurations for mapping each supported toolchains, such as `GCC_ARM`, to their flags, such as `-O3`.
```
I think this sentence is still missing something. We're saying something about the JSON object that represents configurations, but I don't know what we're saying about it. Maybe we just lost a verb somewhere?

SenRamakri replied:
@AnotherButler - I just modified this sentence. Please see if that fixes the overall sentence.

AnotherButler replied:
LGTM, thanks 👍

SenRamakri replied:
@AnotherButler - Great, and Thanks for the quick review.

SenRamakri committed:
Fix the build profile description to capture how the build profiles are implemented

AnotherButler commented:
@theotherjimmy Have your concerns been addressed?

theotherjimmy commented on docs/tools/CLI/build_profiles.md:
```diff
@@ -1,42 +1,42 @@
 <h2 id="build-profiles">Build profiles</h2>
 
-Arm Mbed OS 5 supports three primary build profiles: *develop*, *debug* and *release*. The Online Compiler uses the *develop* profile. When building from Arm Mbed CLI, you can select a profile by adding the `--profile <profile>` flag. You can specify custom user-defined profiles by giving the path to the profile.
+Arm Mbed OS 5 defines three collections of toolchain flags used during the build. These are __build profiles__. These are collections of toolchain flags used during the build. The three build profiles are *develop*, *debug* and *release*. The Mbed Online Compiler uses the *develop* build profile. When building from Arm Mbed CLI, you can select the build configuration by adding the `--profile <`__ ...
```
Remove redundant definition and "build configuration" or "configuration" -> "build profile":
```suggestion
Arm Mbed OS 5 defines three collections of toolchain flags used during the build. These are __build profiles__. The three build profiles are *develop*, *debug* and *release*. The Mbed Online Compiler uses the *develop* build profile. When building from Arm Mbed CLI, you can select the build profile by adding the `--profile <`__build profile__ `name>` flag. You can also specify custom- or user-defined build profiles by giving the path to the JSON file defining the build profile.
```

Further, I'm not sure I like
> by adding the `--profile <`__build profile__ `name>` flag.

I would instead write something like 
> you may select the build profile by passing your desired build profile, by name or path, to the `--profile` argument.

AnotherButler replied:
Apparently, I don't have permissions to push to this repository when using the "commit this suggestion" button, but I can make the changes directly?

Commit:
Clarify paragraph in build_profiles.md

Apply suggestions from comments.

SenRamakri committed:
Modified the introduction and user-defined sections

theotherjimmy commented:
The redundant sentence in the introduction is still present.

theotherjimmy commented on docs/tools/CLI/build_profiles.md:
```diff
@@ -1,42 +1,42 @@
 <h2 id="build-profiles">Build profiles</h2>
 
-Arm Mbed OS 5 supports three primary build profiles: *develop*, *debug* and *release*. The Online Compiler uses the *develop* profile. When building from Arm Mbed CLI, you can select a profile by adding the `--profile <profile>` flag. You can specify custom user-defined profiles by giving the path to the profile.
+Arm Mbed OS 5 defines three collections of toolchain flags used during the build. These are __build profiles__. These are collections of toolchain flags used during the build. The three build profiles are *develop*, *debug* and *release*. The Mbed Online Compiler uses the *develop* build profile. When building from Arm Mbed CLI, you may select the __build profile__ by passing your desired build p ...
```
Remove the redundant sentence, use code format for code.

```suggestion
Arm Mbed OS 5 defines three collections of toolchain flags used during the build. These are __build profiles__.  The three build profiles are *develop*, *debug* and *release*. The Mbed Online Compiler uses the *develop* build profile. When building from Arm Mbed CLI, you may select the __build profile__ by passing your desired build profile, by name or path, to the `--profile` argument.
```

SenRamakri replied:
@AnotherButler - What do you think of @theotherjimmy suggestion? Wanted to check with you as you had that modified little bit.

AnotherButler replied:
@SenRamakri I'm fine with this, but I can't click the "commit suggestion" button. I think there's a bug. Feel free to do so yourself, or I can make the changes to your PR manually.

SenRamakri replied:
@AnotherButler - Thanks for confirming, and I'll make the modifications along with other review comments fixes.

theotherjimmy commented on docs/tools/CLI/build_profiles.md:
```diff
@@ -1,42 +1,42 @@
+#### JSON build profile format
 
-The JSON object that represents a toolchain profile is a dictionary mapping from toolchains, such as `GCC_ARM`, to their flags, such as `-O3`.
+The __build profiles__ are implemented as JSON files with the root object representing toolchain configurations for mapping each supported toolchains, such as `GCC_ARM`, to their flags, such as `-O3`.
```
Build profiles are not implemented. Build profile _are_ JSON files. What does "toolchain configurations mean"? can we just say that they are a JSON object with a key for each toolchain?
```suggestion
The __build profiles__ are JSON files with a root object representing toolchain configurations for mapping each supported toolchains, such as `GCC_ARM`, to their flags, such as `-O3`.
```

or 

```suggestion
__build profiles__ are JSON files that map each supported toolchain, such as `GCC_ARM`, to their flags, such as `-O3`.
```

theotherjimmy commented on docs/tools/CLI/build_profiles.md:
```diff
@@ -1,42 +1,42 @@
+The __build profiles__ are implemented as JSON files with the root object representing toolchain configurations for mapping each supported toolchains, such as `GCC_ARM`, to their flags, such as `-O3`.
 
-The structure is as follows: each toolchain supported by a toolchain profile has a dictionary in the root dictionary. This dictionary contains a mapping from a flag type to a list of flags that should be passed to the corresponding part of the compiler suite.
+Each toolchain to be supported is represented by a child object in the JSON root object. Each child object contains a mapping from a flag type to a list of flags that should be passed to the corresponding part of the compiler suite.
```
"Each toolchain to bu supported" is awkward to me. How about:

```suggestion
In a __build profile__, the root JSON object maps supported toolchains to their child object. Each child object contains a mapping from a flag type to a list of flags that is passed to the corresponding part of the compiler suite.
```

SenRamakri committed:
New fixes for JSON build format section

SenRamakri commented:
@AnotherButler @theotherjimmy - I have new changes committed, please review.

Commit:
Fix typo in build_profiles.md

Add missing letter.

AnotherButler commented:
ping @theotherjimmy

theotherjimmy commented on docs/tools/CLI/build_profiles.md:
```diff
@@ -1,42 +1,41 @@
-
-The structure is as follows: each toolchain supported by a toolchain profile has a dictionary in the root dictionary. This dictionary contains a mapping from a flag type to a list of flags that should be passed to the corresponding part of the compiler suite.
+The __build profiles__ are JSON files with the root object containing key-value pairs for each supported toolchain, such as `GCC_ARM`.
+Each key, representing the toolchain, contains a mapping from a flag type to a list of flags that should be passed to the corresponding part of the compiler suite.
```
The key does not contain a mapping. If anything that's the value. 
```suggestion
Each key is a toolchain name and every value contains a mapping from a flag type to a list of flags that should be passed to the corresponding part of the compiler suite.
```

AnotherButler replied:
Done

Commit:
Fix wording in build_profiles.md

Make suggested change from comments for accuracy.

theotherjimmy commented on docs/tools/CLI/build_profiles.md:
```diff
@@ -1,42 +1,40 @@
+### User-defined build profile
 
-A toolchain or build system profile is a set of flags that is guaranteed to be passed to the underlying compiler suite.
+As mentioned above, the __build profile__ defines the set of flags that is guaranteed to be passed to the underlying compiler suite. You can also create a custom or user-defined __build profile__ using a JSON file acording to the __build profile__ format mentioned in [JSON build profile format](#JSON-build-profile-format).
```
```suggestion
As mentioned above, the __build profile__ defines the set of flags that is passed to the underlying compiler suite. You can also create a custom or user-defined __build profile__ using a JSON file according to the __build profile__ format mentioned in [JSON build profile format](#JSON-build-profile-format).
```

SenRamakri replied:
Good catch.

theotherjimmy commented on docs/tools/CLI/build_profiles.md:
```diff
@@ -1,42 +1,40 @@
+As mentioned above, the __build profile__ defines the set of flags that is guaranteed to be passed to the underlying compiler suite. You can also create a custom or user-defined __build profile__ using a JSON file acording to the __build profile__ format mentioned in [JSON build profile format](#JSON-build-profile-format).
 
-These flags are stored in a JSON file that may be merged with other JSON files of the same structure.
+These flags stored in the JSON file may also be merged with other JSON files of the same structure.
```
```suggestion
These flags stored in the JSON file are merged with other JSON files of the same structure when multiple `--profile` arguments are passed on the command line.
```

SenRamakri committed:
Fix JSON docs section

SenRamakri commented:
@theotherjimmy - Review comments fixed, please re-review.

theotherjimmy reviewed (approved):
:+1:
